domain/model/state/state.go

```diff
@@ -978,6 +978,145 @@ OR uuid IN (SELECT grant_on
 	return rval, nil
 }
 
+// ListModelSummariesForUser lists model summaries of all models the user has
+// access to. If no models are found then a nil slice is returned.
+// TODO(aflynn): 05-08-2024 - The ModelSummary struct includes a machine count,
+// unit count and cpu core count, model status as well as migration status. This
+// information has not yet been migrated over to the relational database. Once
+// it has, it needs to be included here.
+func (s *State) ListModelSummariesForUser(ctx context.Context, userName string) ([]model.ModelSummary, error) {
+	db, err := s.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	q := `
+SELECT    (p.access_type, m.uuid, m.name, m.cloud_name, m.cloud_region_name, 
+          m.model_type, m.cloud_type, m.owner_name, m.cloud_credential_name, 
+          m.cloud_credential_cloud_name, m.cloud_credential_owner_name,
+          m.life, mll.time, m.target_agent_version) AS (&dbModelSummary.*)
+FROM      v_user_auth u
+JOIN      v_permission p ON p.grant_to = u.uuid
+JOIN      v_model m ON m.uuid = p.grant_on
+LEFT JOIN model_last_login mll ON m.uuid = mll.model_uuid
+WHERE     u.removed = false
+AND       u.disabled = false
+AND       u.name = $dbUserName.name
+`
+	name := dbUserName{Name: userName}
+	modelStmt, err := s.Prepare(q, name, dbModelSummary{})
+	if err != nil {
+		return nil, errors.Annotatef(err, "preparing get model statement")
+	}
+
+	controllerInfo := dbController{}
+	controllerUUIDstmt, err := s.Prepare(`
+SELECT &dbController.*
+FROM controller
+`, controllerInfo)
+	if err != nil {
+		return nil, errors.Annotatef(err, "preparing get controller uuid statement")
+	}
+
+	var models []dbModelSummary
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err = tx.Query(ctx, modelStmt, name).GetAll(&models)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return nil
+		} else if err != nil {
+			return errors.Trace(err)
+		}
+
+		err = tx.Query(ctx, controllerUUIDstmt).Get(&controllerInfo)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			// If this happens it something is very wrong.
+			return fmt.Errorf("controller details not found")
+		} else if err != nil {
+			return errors.Trace(err)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, errors.Annotate(err, "getting model summaries for user")
+	}
+
+	modelSummaries := make([]model.ModelSummary, len(models))
+	for i, m := range models {
+		modelSummaries[i], err = m.decodeModelSummary(controllerInfo)
+		if err != nil {
+			return nil, errors.Annotate(err, "getting model summaries for user")
```

Comment:
We should add more context here including the user we are doing this on behalf of.

Reply:
See Simons comment about removing the user name from here. It can be a security issue. I had the username here before.
https://github.com/juju/juju/pull/17861#discussion_r1704006531